sentinelhub/time_utils.py

```diff
@@ -147,34 +161,80 @@ def parse_time_interval(time):
     :param time: An input time
     :type time: str or datetime.datetime or (str, str) or (datetime.datetime, datetime.datetime)
     :return: interval of start and end date of the form `YYYY-MM-DDThh:mm:ss`
-    :rtype: (str, str)
+    :rtype: (datetime.datetime, datetime.datetime)
     :raises: ValueError
     """
     if time == SHConstants.LATEST:
-        date_interval = '1985-01-01', get_current_date()
-    elif isinstance(time, (str, datetime.date)):
-        parsed_time = parse_time(time)
+        return dt.datetime(year=1985, month=1, day=1), dt.datetime.now()
+
+    if isinstance(time, (str, dt.date)):
+        parsed_time = parse_time(time, **kwargs)
         date_interval = parsed_time, parsed_time
     elif isinstance(time, (tuple, list)) and len(time) == 2:
-        date_interval = parse_time(time[0]), parse_time(time[1])
+        date_interval = parse_time(time[0], **kwargs), parse_time(time[1], **kwargs)
     else:
         raise ValueError('Time must be a string/datetime object or tuple/list of 2 strings/datetime objects')
 
-    if 'T' not in date_interval[0]:
-        date_interval = date_interval[0] + 'T00:00:00', date_interval[1]
-    if 'T' not in date_interval[1]:
-        date_interval = date_interval[0], date_interval[1] + 'T23:59:59'
+    start_time, end_time = date_interval
+    if not isinstance(start_time, dt.datetime):
+        start_time = date_to_datetime(start_time)
+    if not isinstance(end_time, dt.datetime):
+        end_time = date_to_datetime(end_time, dt.time(hour=23, minute=59, second=59))
 
-    if date_interval[1] < date_interval[0]:
+    if start_time > end_time:
         raise ValueError('Start of time interval is larger than end of time interval')
 
-    return date_interval
+    return start_time, end_time
+
+
+def serialize_time(timestamp_input, *, use_tz=False):
```

Comment:
I am not too keen on differences between parsing time and serializing time, namely, parsing works on single time while serializing can do a tuple. 

If that is the case, I'd prefer that `parse_time` (or perhaps `parse_datetime`) also works on a tuple. One can thus get rid of `parse_time_interval`. Perhaps it would be prudent to limit these functions to work on a single datetime or on a pair (to mimic time_interval). This would make the methods much more similar.

Reply:
I thought about this but the problem is that `parse_time_interval` is combining information about the start and end time while `serialize_time` is serializing each time independently. I.e.:

- `parse_time_interval(time_input)` is in most cases not the same as `parse_time(time_input[0]), parse_time(time_input[1])`
- `serialize_time((start_time, end_time))` is the same as `serialize_time(start_time), serialize_time(end_time)`.

Because of that, I don't think `parse_time` and `parse_time_interval` can be joined nicely.

Reply:
The fact that `parse_time_interval(time_input)` is not equal to `parse_time(time_input[0]), parse_time(time_input[1])` is the biggest issue I have. 

I know it allows nearly anything to be parsed, but that is a way to hiding explicitly providing `time_interval` from two dates/times (e.g., `start_time` and `end_time`). The case where `end_time == start_time`, is just a particular case. 

Validation of `time_interval` should be done elsewhere.

Reply:
I see `parse_time_interval` as a transformation of various "loose" time definitions into a strict time interval definition with 2 `datetime.datetime` objects. Sometimes this is quite convenient.

For example, let's say you want an image for `2021-02-24`. With `parse_time_interval` you can make a Process API request with `time='2021-02-24'`. Without it you would have to write `time=('2021-02-24T00:00:00Z', '2021-02-24T23:59:59Z')`. For prototyping with sh-py, I think this soon becomes a bit annoying from the user's point of view.

Also, if we go further changing how `parse_time_interval` works I think this will break quite a few things in our codebase as well.